storage/innobase/handler/ha_innodb.cc

```diff
@@ -18668,103 +18668,91 @@ void lock_wait_wsrep_kill(trx_t *bf_trx, ulong thd_id, trx_id_t trx_id)
-	wsrep_thd_LOCK(victim_thd);
-	trx_t* victim_trx= thd_to_trx(victim_thd);
-	wsrep_thd_UNLOCK(victim_thd);
-
-	WSREP_DEBUG("abort transaction: BF: %s victim: %s victim conf: %s",
-			wsrep_thd_query(bf_thd),
-			wsrep_thd_query(victim_thd),
-			wsrep_thd_transaction_state_str(victim_thd));
-
-	if (victim_trx) {
-		victim_trx->lock.set_wsrep_victim();
-
-		wsrep_thd_LOCK(victim_thd);
-		bool aborting= !wsrep_thd_set_wsrep_aborter(bf_thd, victim_thd);
-		wsrep_thd_UNLOCK(victim_thd);
-		if (aborting) {
-			DEBUG_SYNC(bf_thd, "before_wsrep_thd_abort");
-			DBUG_EXECUTE_IF("sync.before_wsrep_thd_abort",
-					 {
-					   const char act[]=
-					     "now "
-					     "SIGNAL sync.before_wsrep_thd_abort_reached "
-					     "WAIT_FOR signal.before_wsrep_thd_abort";
-					   DBUG_ASSERT(!debug_sync_set_action(bf_thd,
-									      STRING_WITH_LEN(act)));
-					 };);
-			wsrep_thd_bf_abort(bf_thd, victim_thd, signal);
-		}
-	} else {
-		DBUG_EXECUTE_IF("sync.before_wsrep_thd_abort",
-				 {
-				   const char act[]=
-				     "now "
-				     "SIGNAL sync.before_wsrep_thd_abort_reached "
-				     "WAIT_FOR signal.before_wsrep_thd_abort";
-				   DBUG_ASSERT(!debug_sync_set_action(bf_thd,
-								      STRING_WITH_LEN(act)));
-				 };);
-		wsrep_thd_bf_abort(bf_thd, victim_thd, signal);
-	}
-
-	wsrep_thd_kill_UNLOCK(victim_thd);
-	DBUG_VOID_RETURN;
+static void wsrep_abort_transaction(handlerton *, THD *bf_thd, THD *victim_thd,
+                                    my_bool signal)
+{
+  DBUG_ENTER("wsrep_abort_transaction");
+  ut_ad(bf_thd);
+  ut_ad(victim_thd);
+
+  trx_t *victim_trx= thd_to_trx(victim_thd);
+
+  WSREP_DEBUG("abort transaction: BF: %s victim: %s victim conf: %s",
+              wsrep_thd_query(bf_thd), wsrep_thd_query(victim_thd),
+              wsrep_thd_transaction_state_str(victim_thd));
+
+  if (!victim_trx)
+  {
+    WSREP_DEBUG("abort transaction: victim did not exist");
+    DBUG_VOID_RETURN;
+  }
+
+  victim_trx->mutex_lock();
+
+#ifdef ENABLED_DEBUG_SYNC
+  if (victim_trx->state == TRX_STATE_NOT_STARTED)
+  {
+    DBUG_EXECUTE_IF("sync.wsrep_abort_transaction_read_only", {
+      const char act[]=
+          "now "
+          "SIGNAL sync.wsrep_abort_transaction_read_only_reached "
+          "WAIT_FOR signal.wsrep_abort_transaction_read_only";
+      DBUG_ASSERT(!debug_sync_set_action(bf_thd, STRING_WITH_LEN(act)));
+    };);
+  }
+#endif /* ENABLED_DEBUG_SYNC */
+
+  if (victim_trx->state == TRX_STATE_ACTIVE)
+  {
+    DEBUG_SYNC(bf_thd, "before_wsrep_thd_abort");
+    DBUG_EXECUTE_IF("sync.before_wsrep_thd_abort", {
+      const char act[]= "now "
+                        "SIGNAL sync.before_wsrep_thd_abort_reached "
+                        "WAIT_FOR signal.before_wsrep_thd_abort";
+      DBUG_ASSERT(!debug_sync_set_action(bf_thd, STRING_WITH_LEN(act)));
+    };);
+    victim_trx->lock.set_wsrep_victim();
+    victim_trx->mutex_unlock();
+    lock_sys.cancel_lock_wait_for_trx(victim_trx);
```

Comment:
Is it absolutely necessary to enter a `DEBUG_SYNC` wait while holding `victim_trx->mutex`?

There is no guarantee that the transaction exists at the time `lock_sys_t::cancel_lock_wait_for_trx(trx_t *)` is called. The transaction commit could be blocked in `trx_t::commit_state()` until the moment `victim_trx->mutex_unlock()` is called. After that point, the call to `wsrep_commit_ordered()` might cause some synchronization; I am not familiar with that code. A little after that call, `trx_t::commit_cleanup()` can free the transaction object back to the transaction pool, and the memory can be reused for an unrelated transaction. In other words, the call `lock_sys.cancel_lock_wait_for_trx(victim_trx);` here may kill an innocent transaction if the intended `victim_trx` was committed right before the call.

I think that safe operation here needs to follow a similar locking pattern as `lock_sys_t::cancel_lock_wait_for_trx(trx_t*)`.

There should also be a comment that explains what the intended impact is on transactions that are not blocked in `lock_wait()` but waiting for something outside InnoDB, such as the next SQL statement to execute, or metadata lock on some table.

Reply:
My suggestion to this: Cache victim_trx->id while we hold victim_trx mutex. Then either extend ock_sys_t::cancel_lock_wait_for_trx(trx_t*) or introduce new wsrep version where after we have acquired lock_sys latch and lock_sys wait mutex we validate victim_trx id and state, we could even use trx_sys::find() to avoid referencing victim_trx before we know it is valid.

Reply:
Yes, storing the `trx_t::id` should be safe. Transaction identifiers are not reusable and InnoDB does not support wrap-around. At a rate of 10⁶ transactions per second, the space of 2⁴⁸ identifiers (actually 2⁴⁷ because a second one is consumed on transaction commit) would be consumed in 4 years and 5 months.